pkg/apply/destroyer.go

```diff
@@ -68,7 +72,9 @@ func (d *Destroyer) Run(inv inventory.InventoryInfo, option *DestroyerOption) <-
 
 	go func() {
 		defer close(ch)
-		d.invClient.SetDryRunStrategy(d.DryRunStrategy)
+		// TODO: We should see if we can avoid having the dry-run strategy
+		// as a field on the inventory client.
```

Comment:
I was looking at that just yesterday! It's very unfortunate that the client is mutable and hence it's not safe for concurrent use. And because of that the `Applier` is also not thread safe. I wanted to ask - would it make sense to refactor the constructor+`Initialize()` into a more traditional pattern of factory+immutable object i.e. remove `Initialize()` and remove mutability. Then the "dry run" strategy can be passed to the client on construction and the client is immutable. Client construction is done in Applier as needed, without any mutation. That is what I came up with yesterday, maybe there is a better way. WDYT?

I'm happy to make the refactor and to update `kpt` too to acomodate the breaking changes.

Reply:
Found it https://github.com/GoogleContainerTools/kpt/issues/1462

Reply:
Yeah, it'd be better as a parameter to client's methods than a constructor parameter.

Reply:
Yeah, we want the Applier and the Destroyer to be stateless and immutable so users can create it once and then use it multiple times. This also includes the InventoryClient. I'm also hoping we eventually can simplify the interface of the InventoryClient as it currently has 7-8 functions that seems somewhat redundant.
The `Initialize` functions in both the Applier and the Destroyer are remains from when we used the kubectl libraries in a different way, and I agree we want to avoid those if possible. 

If you want to refactor this, that would be great! The kpt team will take care of any updates needed in kpt when we pull in the changes, so no need to worry about that unless you want to contribute.